File: credit-engine-api/src/main/java/com/credit/engine/application/dto/receivable/ReceivableRequest.java
package com.credit.engine.application.dto.receivable;

import com.credit.engine.domain.model.receivable.ReceivableType;
import jakarta.validation.constraints.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.UUID;

public record ReceivableRequest(
        @NotNull(message = "O identificador do cedente é obrigatório.")
        UUID assignorId,

        @NotNull(message = "O tipo do título é obrigatório.")
        ReceivableType type,

        @NotBlank(message = "O número do documento é obrigatório.")
        @Size(max = 50, message = "O número do documento deve conter no máximo 50 caracteres.")
        String documentNumber,

        @NotNull(message = "O valor de face é obrigatório.")
        @DecimalMin(value = "0.0", inclusive = false, message = "O valor de face deve ser maior que zero.")
        @Digits(integer = 14, fraction = 4, message = "O valor de face deve conter no máximo 14 dígitos inteiros e 4 casas decimais.")
        BigDecimal faceValue,

        @NotBlank(message = "O código da moeda é obrigatório.")
        @Size(min = 3, max = 3, message = "O código da moeda deve possuir exatamente 3 letras (ex: BRL, USD).")
        String currencyCode,

        @NotNull(message = "A data de vencimento é obrigatória.")
        @FutureOrPresent(message = "A data de vencimento não pode estar no passado.")
        LocalDate dueDate
) {
}

File: credit-engine-ui/src/services/receivable/dto/ReceivableRequest.ts
import ValidationUtils from '../../../utils/ValidationUtils.ts';
import type { ReceivableType } from '../ReceivableType.ts';

export default interface ReceivableRequest {
    assignorId: string;
    type: ReceivableType | '';
    documentNumber: string;
    faceValue: number | null;
    currencyCode: string;
    dueDate: Date | null;
}

export const validateReceivableRequest = (request: ReceivableRequest): { message: string }[] => {
    const errors: { message: string }[] = [];

    if (ValidationUtils.isBlank(request.assignorId))
        errors.push({ message: 'Selecione um cedente antes de simular.' });

    if (ValidationUtils.isBlank(request.type))
        errors.push({ message: 'O tipo do recebível é obrigatório.' });

    if (ValidationUtils.isBlank(request.documentNumber))
        errors.push({ message: 'O numero do documento do titulo é obrigatório.' });

    if (request.faceValue === null || request.faceValue <= 0)
        errors.push({ message: 'O valor de face deve ser maior que zero.' });

    if (ValidationUtils.isBlank(request.currencyCode))
        errors.push({ message: 'A moeda do titulo é obrigatória.' });

    if (!request.dueDate)
        errors.push({ message: 'A data de vencimento é obrigatória.' });

    return errors;
};
